### PR TITLE
persist: Migrate Stats entirely over to Columnar V2

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR migrates the stats handling code entirely over to Columnar V2 and removes any dependency on the `Data` trait, this unblocks the deletion of Columnar V1!

Note: We're already exercising the Columnar V2 stats handling code in Production, this PR should just purely be cleanup.

### Motivation

Delete the `Data` trait

### Tips for reviewer

There are a few logically separate changes in this PR that I split into separate commits:

1. Change the `interesting_datums(...)` function in `mz_repr` to return only valid JSON numbers. Previously it returned Infinity and NaN which are not valid in JSON.
2. Update `PartBuilder2` so the `time` and `diff` values are generic over `Codec64`, this brings it inline with `PartBuilder`
3. ⭐⭐⭐ **The most important commit**, it refactors the calculation of a `ResultSpec` from `PartStats` to use only ColumnarV2 types
4. Refactor of other small parts of the code base to remove the dependency of the `Data` trait from statistics.
5. Regenerate the statistics stability snapshot

The statistics stability snapshot changed because we encode UUID, Interval, and Numerics differently now. I'm also `cargo stress`-ing the proptests for stats roundtripping.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

**Review by jkosh44 — COMMENTED** [review]
Adapter and txn-wal changes LGTM (I'll leave the actual approval to persist)

**Review by bkirwi — APPROVED** [review]
Checking my understanding -
- This diff comes with a large change to the snaps. This is because the old stats code was still in effect for those tests until this change. (Though we've already moved staging etc. over to the new stuff.)
- The reason that kind of diff is okay is because the new stats stuff handles both old and new formats - for example, numeric can be decoded from either atomic stats or the new fixed-size stats.

Assuming that checks out... 👍!

File: src/storage-types/src/stats.rs [hunk 1]
```diff
@@ -152,26 +102,10 @@ impl RelationPartStats<'_> {
 
     pub fn col_stats<'a>(&'a self, id: usize, arena: &'a RowArena) -> ResultSpec<'a> {
         // While we migrate to our new columnar data setup, check both old and new versions.
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/storage-types/src/stats.rs]: Good catch, removed!

File: src/persist-types/src/stats.rs [hunk 2]
```diff
@@ -89,6 +89,64 @@ impl ColumnarStats {
...
+
+    /// Helper method to "downcast" to stats of type `T`.
+    fn try_as_stats<'a, T, F>(&'a self, map: F) -> Result<T, anyhow::Error>
+    where
+        F: FnOnce(&'a ColumnStatKinds) -> Result<T, anyhow::Error>,
+    {
+        let inner = map(&self.values)?;
+        match self.nulls {
+            Some(_) => Err(anyhow::anyhow!(
+                "expected non-nullable PrimitiveStats<String>, found nullable"
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/persist-types/src/stats.rs]: Ahhh thanks for catching this, copy + paste error, fixed!

File: src/persist-types/src/stats.rs [hunk 3]
```diff
@@ -89,6 +89,64 @@ impl ColumnarStats {
...
+            other => anyhow::bail!("expected BytesStats found {other:?}"),
+        })
+    }
+
+    /// Tries to "downcast" this instance of [`ColumnarStats`] to a [`PrimitiveStats<String>`]
+    /// if the inner statistics are nullable and for a structured column.
+    pub fn try_as_string(&self) -> Result<&PrimitiveStats<String>, anyhow::Error> {
+        self.try_as_stats(|values| match values {
+            ColumnStatKinds::Primitive(PrimitiveStatsVariants::String(inner)) => Ok(inner),
+            other => anyhow::bail!("expected StructStats found {other:?}"),
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/persist-types/src/stats.rs]: Similar copy+paste issue, fixed!

**Pull request merged**